app/src/main/java/org/thoughtcrime/securesms/crypto/ModernEncryptingPartOutputStream.java

```diff
@@ -0,0 +1,61 @@
+package org.thoughtcrime.securesms.crypto;
+
+
+import android.util.Pair;
+
+import androidx.annotation.NonNull;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+
+import javax.crypto.Cipher;
+import javax.crypto.CipherOutputStream;
+import javax.crypto.Mac;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * Constructs an OutputStream that encrypts data written to it with the AttachmentSecret provided.
+ *
+ * The on-disk format is very simple, and intentionally no longer includes authentication.
+ */
+public class ModernEncryptingPartOutputStream {
+
+  public static Pair<byte[], OutputStream> createFor(@NonNull AttachmentSecret attachmentSecret, @NonNull File file, boolean inline)
+      throws IOException
+  {
+    byte[] random = new byte[32];
+    new SecureRandom().nextBytes(random);
+
+    try {
+      Mac mac = Mac.getInstance("HmacSHA256");
+      mac.init(new SecretKeySpec(attachmentSecret.getModernKey(), "HmacSHA256"));
+
+      FileOutputStream fileOutputStream = new FileOutputStream(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *fileOutputStream*. Currently, there are execution paths that do not contain closure statements, for example, when Cipher.init throws an exception. Ensure *fileOutputStream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).